README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![natverse](https://img.shields.io/badge/natverse-Part%20of%20the%20natverse-a241b6)](https://natverse.github.io)
[![Travis build
status](https://travis-ci.org/natverse/mouselightr.svg?branch=master)](https://travis-ci.org/natverse/mouselightr)
[![Codecov test
coverage](https://codecov.io/gh/natverse/mouselightr/branch/master/graph/badge.svg)](https://codecov.io/gh/natverse/mouselightr?branch=master)
[![Docs](https://img.shields.io/badge/docs-100%25-brightgreen.svg)](http://natverse.github.io/mouselightr/reference/)
<img src="man/figures/logo.svg" align="right" height="139" />

Mouselightr
===========

The goal of *mouselightr* is to provide R client utilities for
interacting with the [MouseLight](http://mouselight.janelia.org/) at
[Janelia Research Campus](https://www.janelia.org/), which has
[successfully](https://www.cell.com/neuron/pdfExtended/S0896-6273(19)30391-5)
acquired and registered almost \~1,000 neurons from the mouse into a
standard, annotated template space. MouseLight investigators have
acquired sub-micron single neuron reconstructions from two-photon
microscopy datasets of whole brainsusing modified tissue clearing
techniques, and have generated a template brain with defined
sub-volumes. Using this R package in concert with the
[natverse](https://github.com/natverse/natverse) ecosystem of
neuroanatomy tools is highly recommended.

Installation
------------

Firstly, you will need R, R Studio and X Quartz as well as nat and its
dependencies. For detailed installation instructions for all this, see
[here](https://natverse.github.io/nat/articles/Installation.html). It
should not take too long at all. Then:

``` r
# install
if (!require("devtools")) install.packages("devtools")
devtools::install_github("natverse/mouselightr")

# use 
library(mouselightr)
```

Done!

Key Functions
-------------

Now we can have a look at what is available, here are some of the key
functions. Their help details examples of their use. You can summon the
help in RStudio using `?` followed by the function name.

``` r
# And how can I download and read neurons?
?mouselight_read_neurons()

# Get useful information on the neurons available
?mouselight_neuron_info()

# Get a 3D neuropil-subdivided brain model
?mouselight_read_brain # Get 3D neuropil-subdivided brain models for those brainspaces
```

Example
-------

<img src="man/figures/mouselight_neuropils.png" />

Let’s also have a look at an example pulling neurons and brain meshes
from the [MouseLight neuron browser](https://mouselightr.neuro.mpg.de).
Excitingly, all the data is in a single standard template space!

``` r
## First we can quickly just plot the outer mesh for the brain
outline = mouselight_read_brain(type = "outline")
plot3d(outline, col = "pink", alpha = 0.3)

## This is cool, but maybe what we really want are its sub-divisions.
mousebrain = mouselight_read_brain(type = "brain_areas")
clear3d()
plot3d(mousebrain)
### This takes a long time the first time you call this function per session
### .obj files are saved in a temporary folders
### You could save this brain locally, to use quickly in future.
#### save(mousebrain,"YOUR_PATH/mouselight_brain.rda")

## What brain regions are on offer?
print(mousebrain$neuropil_full_names)

## Or if we want more information, we can get it like this:
mbr = mouselight_brain_region_info()
View(mouselight_brain_region_info)

## Perhaps we want to plot just the amygdala?
### To do this we can do
amygdala.codes = mousebrain$RegionList[grepl("amygdala",mousebrain$neuropil_full_names,
ignore.case = TRUE)]
plot3d(outline, col = "pink", alpha = 0.1)
plot3d(subset(mousebrain, amygdala.codes), alpha = 0.5)

## So now we want some neurons
## What neurons data is available?
ndf=mouselight_neuron_info()
#### How many tracings per neurons?
table(table(ndf$neuron.id))
#### This is because many 'neurons' have a separate axon and dendrite skeleton 

## We can download all of these neurons, and their meta-data
### Typically two tracings, and axon and a dendrite, per neuron
mlns = mouselight_read_neurons(ndf$tracing.id, meta = TRUE)

## Let's read in all the amygdalal neurons
### Since each 
in.amyg = mouselight_nodes_in_region(mlns, brain.areas = amygdala.codes, labels = NULL)
amyg.ids = names(in.amyg)[in.amyg>0]
amyg.neurons = mlns[amyg.ids]

## And plot!
plot3d(amyg.neurons)
### Quite wide-ranging!
```

<img src="man/figures/mouselight_neurons.png" />

Acknowledging the data and tools
--------------------------------

Any work that uses data from this package should cite

**Winnubst, Johan, Erhan Bas, Tiago A. Ferreira, Zhuhao Wu, Michael N.
Economo, Patrick Edson, Ben J. Arthur, et al.** 2019. *Reconstruction of
1,000 Projection Neurons Reveals New Cell Types and Organization of
Long-Range Connectivity in the Mouse Brain.* bioRxiv.
<a href="https://doi.org/10.1101/537233" class="uri">https://doi.org/10.1101/537233</a>.

**Economo, Michael N., Nathan G. Clack, Luke D. Lavis, Charles R.
Gerfen, Karel Svoboda, Eugene W. Myers, and Jayaram Chandrashekar.**
2016. *A Platform for Brain-Wide Imaging and Reconstruction of
Individual Neurons.* eLife 5 (January): e10566.

This package was created by [Alexander Shakeel
Bates](https://scholar.google.com/citations?user=BOVTiXIAAAAJ&hl=en) and
[Dr. Gregory Jefferis](https://en.wikipedia.org/wiki/Gregory_Jefferis).
You can cite this package as:

``` r
citation(package = "mouselightr")
```

**Bates AS, Jefferis GSXE** (2019). *mouselightr: R client utilities for
interacting with the MouseLight project.* **R package** version 0.1.0.
<a href="https://github.com/natverse/mouselightr" class="uri">https://github.com/natverse/mouselightr</a>


NEWS.md
# mouselightr 0.1.0

* complete rename of package to mouselightr

# nat.mouselight 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/info.R
#' @title Get information about the mouse brain compartments in the MouseLight project
#'
#' @description Query MouseLight's GraphQL API to retreive information about all of the brain volumes in the MouseLight project. These
#' brain regions can be sub-volumes of other brain regions. An \code{igraph} object can be generated to explore these dependencies.
#' @param ... methods passed to \code{mouselight_fetch}
#' @return Using \code{mouselight_brain_graph} will return an object of class \code{igraph} for use with the \code{igraph} R package.
#' Using \code{mouselight_brain_region_info}, a \code{data.frame} with the following values is returned:
#'\itemize{
#'  \item 	\emph{acronym}	 short name for a brain region
#'  \item 	\emph{safeName}	 long, easily computer-readble name for a brain region
#'  \item 	\emph{structureId}	 unique ID for brain region
#'  \item 	\emph{parentStructureId}	 ID for a larger brain region, to which this brain region belongs
#'  \item 	\emph{structureIdPath}	 a path of structure IDs to the 'root', the whole brain
#'}
#' @seealso \code{\link{mouselight_read_brain}}, \code{\link{mouselight_read_neurons}}
#' @examples
#' \dontrun{
#' ## First we need to download all of the neurons
#' mbr = mouselight_brain_region_info()
#'
#' ## Leet's have a look at this data
#' View(mouselight_brain_region_info)
#'
#' ## Some brain regions are actually a subcompartment of others in this set.
#' ### Hmm, so what would be good is if we could see those correspondences
#' g = mouselight_brain_graph()
#'
#' ## Let's plot these i na way we can easily see!
#' library(igraph)
#' f = igraph::layout_with_fr(g)
#' plot(g, layout = f)
#'
#' ## There are a lot of nodes here, we might want to move them about
#' tkplot(g, layout = f)
#'
#' }
#' @inherit mouselight_read_brain references
#' @export
#' @rdname mouselight_brain_info
mouselight_brain_region_info <- function(...){
  body = list(
    query = "{\n
    brainAreas {\n
    acronym\n
    safeName\n
    structureId\n
    parentStructureId\n
    structureIdPath\n
    id\n
    }\n
}",
    variables = list(),
    operationName = NULL
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res = mouselight_fetch(path = "graphql",
                 body = bodyj,
                 parse.json = TRUE,
                 simplifyVector=FALSE,
                 include_headers = FALSE,
                 config = httr::content_type_json(),
                 encode='raw',
                 ...)
  df = data.frame(do.call(rbind, lapply(res$data$brainAreas, function(x) unlist(nullToNA(x)))))
  rownames(df) = df$structureId
  df$parentStructureId[is.na(df$parentStructureId)] = -1
  df
}

#' @export
#' @rdname mouselight_brain_info
mouselight_brain_graph <- function(...){
  bi = mouselight_brain_region_info(...)
  bi = bi[!apply(bi, 1, function(x) sum(is.na(x))>0),]
  vertices <- data.frame(name = bi$structureId,
                         label= bi$acronym)
  relations <- data.frame(from = bi$parentStructureId,
                          to= bi$structureId)
  relations = relations[relations$from != -1 & relations$to != -1,]
  igraph::graph_from_data_frame(relations, directed=TRUE, vertices=vertices)
}

#' @title Get information about the MouseLight project's GraphQL API
#'
#' @description See the API version and the number of neurons hosted by the project.
#' @param ... methods passed to \code{mouselight_fetch}
#' @return a named vector of values
#' @seealso \code{\link{mouselight_read_brain}}, \code{\link{mouselight_read_neurons}}
#' @examples
#' \donttest{
#' mouselight_api()
#' }
#' @inherit mouselight_read_brain references
#' @export
#' @rdname mouselight_info
mouselight_api <- function(...){
  body = list(
    query = "{\n
    systemSettings {\n
    apiVersion\n
    apiRelease\n
    neuronCount\n
    }\n
    }",
    variables = list(),
    operationName = NULL
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res = mouselight_fetch(path = "graphql",
                 body = bodyj,
                 parse.json = TRUE,
                 simplifyVector=FALSE,
                 include_headers = FALSE,
                 config = httr::content_type_json(),
                 encode='raw')
  unlist(res$data$systemSettings)
}

# hidden
structureIdentifiers <- function(){
  body = list(
    query = "{\n
    structureIdentifiers {\n
    id\n
    name\n
    value\n
    }\n
  }",
    variables = list(),
    operationName = NULL
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res = mouselight_fetch(path = "graphql",
                 body = bodyj,
                 parse.json = TRUE,
                 simplifyVector=FALSE,
                 include_headers = FALSE,
                 config = httr::content_type_json(),
                 encode='raw')
  df = do.call(rbind,res$data$structureIdentifiers)
  df = as.data.frame(t(apply(df, 1, unlist)))
  df$value = as.integer(df$value)
  df$`__typename` = "StructureIdentifier"
  rownames(df) = df$structureId
  df
}

# hidden
brainAreaIds <- function(){
  body = list(
    query = "{\n
    brainAreas {\n
    id\n
    acronym\n
    }\n
    }",
    variables = list(),
    operationName = NULL
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res = mouselight_fetch(path = "graphql",
                         body = bodyj,
                         parse.json = TRUE,
                         simplifyVector=FALSE,
                         include_headers = FALSE,
                         config = httr::content_type_json(),
                         encode='raw')
  df = do.call(rbind,res$data$brainAreas)
  df = as.data.frame(t(apply(df, 1, unlist)))
  df$`__typename` = "brainAreaIds"
  rownames(df) = df$structureId
  df
}


R/fetch.R
# hidden
mouselight_fetch <- function(path,
                    body = NULL,
                    parse.json = TRUE,
                    simplifyVector=FALSE,
                    include_headers = FALSE, ...){
  path = gsub("\\/$|^\\/","",path)
  req <-
    if (is.null(body)) {
      httr::GET(url = file.path(mouselight_url(path), path, fsep = "/"), ...)
    }else {
      httr::POST(url = file.path(mouselight_url(path), path, fsep = "/"),
                 body = body, ...)
    }
  httr::stop_for_status(req)
  if (parse.json) {
    parsed = mouselight_parse_json(req, simplifyVector = simplifyVector, raw = FALSE)
    mouselight_error_check(parsed)
    if (include_headers) {
      fields_to_include = c("url", "headers")
      attributes(parsed) = c(attributes(parsed), req[fields_to_include])
    }
    parsed
  }
  else req
}

# hidden
mouselight_parse_json <- function (req, simplifyVector = FALSE, raw = TRUE, ...) {
  if(raw){
    text <- rawToChar(req$content)
  }else{
    text <- httr::content(req, as = "text", encoding = "UTF-8")
  }
  if (identical(text, "")){
    warning("No output to parse", call. = FALSE)
    return(NULL)
  }
  p = tryCatch(jsonlite::fromJSON(text, simplifyVector = simplifyVector, ...), error = function(e) NULL)
  if(is.null(p)){
    warning("error parsing JSON")
  }
  nullToNA(p)
}

# hidden
mouselight_error_check <- function(x){
    err_fields = c("error", "message")
    if (sum(names(x) %in% err_fields)>1) {
      stop(x$error, ": ", x$message)
    }
    NULL
}


R/query.R
#' Query available neurons from MouseLight site
#'
#' @description Query tracings for neurons, and/or neuron soma locations, from the MouseLight Neuron Browser.
#' Each neuron may have multiple tracgings, typically one for its
#' axon and another for its dendrite.
#' @param simplify whether to simplify the query result to a simple data.frame
#'   with one row per tracing (See details).
#' @return data.frame describing the available neurons. See details for the
#'   implication of the \code{simplify} argument.
#' @details Mouselight neurons typically consist of two separate tracings, one
#'   for the axon and one for the dendrites. When \code{simplify=TRUE} (the
#'   default) the output is a \emph{flattened} data frame with one row for every
#'   tracing. When \code{simplify=FALSE} there is one row per neuron, but the
#'   data.frame contains nested data frames with per tracing information.
#'
#'   The identifier which you will need to download a neuron with
#'   \code{\link{mouselight_read_neurons}} is stored in column
#'
#'
#' @examples
#' \donttest{
#' ndf=mouselight_neuron_info()
#' # How many tracings per neurons?
#' table(table(ndf$neuron.id))
#'
#' # In what brain regions do neurons have their end points?
#' br_search = mouselight_neurons_by_brain_region("dendrite")
#' View(br_search)
#'
#' }
#' @seealso \code{\link{mouselight_read_brain}}, \code{\link{mouselight_brain_region_info}}
#' @export
#' @rdname mouselight_neuron_info
mouselight_neuron_info <- function(simplify=TRUE) {

  body = list(
    operationName = "SearchNeurons",
    variables = list(context = list(
      scope = 6L,
      nonce = "cjyo7xu7k00033h5yrj9jfpoy",
      predicates = list(
        list(
          predicateType = 3L,
          tracingIdsOrDOIs = list("1"),
          tracingIdsOrDOIsExactMatch = FALSE,
          tracingStructureIds = list("68e76074-1777-42b6-bbf9-93a6a5f02fa4"),
          nodeStructureIds = list("c37953e1-a1e9-4b9a-847e-08d9566ced65"),
          operatorId = NULL,
          amount = 0L,
          brainAreaIds = list(),
          arbCenter = list(x = NULL, y = NULL, z = NULL),
          arbSize = NULL,
          invert = FALSE,
          composition = 3L
        )
      )
    )),
    query= "query SearchNeurons($context: SearchContext) {\n  searchNeurons(context: $context) {\n    totalCount\n    queryTime\n    nonce\n    \n    neurons {\n      id\n      idString\n      tracings {\n        id\n        tracingStructure {\n          id\n          name\n          value\n          __typename\n        }\n        soma {\n          id\n          x\n          y\n          z\n          radius\n          parentNumber\n          sampleNumber\n          brainAreaId\n          structureIdentifierId\n          __typename\n        }\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n"
  )

  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res=httr::POST(url = mouselight_url('graphql'), body = bodyj, httr::content_type_json(), encode='raw')
  httr::stop_for_status(res)
  raw_res=httr::content(res, as='text', type='application/json', encoding = 'utf8')
  parsed_res=jsonlite::fromJSON(raw_res, simplifyVector = T)
  parsed_res=parsed_res$data[[1]]
  if(is.null(parsed_res$neurons))
    stop("invalid return structure!")
  if(!isTRUE(nneurons <- parsed_res$totalCount>0))
    stop("No neurons returned!")
  parsed_res <- parsed_res$neurons
  if(simplify) {
    stopifnot(isTRUE(all(
      names(parsed_res) %in% c("id", "idString", "brainArea", "tracings", "__typename", "sample")
    )))

    # check soma/tracingStructure columns have same number of rows
    if(!isTRUE(all(sapply(parsed_res$tracings, function(x) nrow(x$tracingStructure)==nrow(x$soma)))))
      stop("Parse Error!",
           "Unexpected mismatch between tracingStructure and soma elements of tracings column!")

    # remove nested tracings data.frames
    neurondf=parsed_res[names(parsed_res)!='tracings']
    # flatten brainArea data.frame
    neurondf=jsonlite::flatten(neurondf)
    # first id is for the neuron (which usually has 2 associated tracings)
    names(neurondf)[1]='neuron.id'

    # flatten nested tracings data.frames
    tracedflist=sapply(parsed_res[['tracings']], jsonlite::flatten, simplify = F)
    names(tracedflist)=neurondf[['neuron.id']]
    tracedf=dplyr::bind_rows(tracedflist, .id='neuron.id')
    names(tracedf)[names(tracedf)=='id']='tracing.id'
    # join neuron and tracings data.frames, one row for each tracing
    parsed_res <- dplyr::right_join(neurondf, tracedf, by='neuron.id')
  }
  names(parsed_res) <- sub("__", "", names(parsed_res))
  parsed_res
}

#' @export
#' @rdname mouselight_neuron_info
mouselight_neuron_somata <- function() {
  query <-  "query SearchNeurons($context: SearchContext) {\n  searchNeurons(context: $context) {\n    neurons {\n      id\n      idString\n tracings {\n soma {\n          x\n          y\n          z\n          radius\n}\n}\n}\n}}\n\n"
  body = list(
    operationName = "SearchNeurons",
    variables = list(context = list(
      scope = 6L,
      nonce = "cjyo7xu7k00033h5yrj9jfpoy",
      predicates = list(
        list(
          predicateType = 3L,
          tracingIdsOrDOIs = list("67f04b64-80ed-4dfc-a519-7cac0605a84f"),
          tracingIdsOrDOIsExactMatch = FALSE,
          tracingStructureIds = list("68e76074-1777-42b6-bbf9-93a6a5f02fa4"),
          nodeStructureIds = list("c37953e1-a1e9-4b9a-847e-08d9566ced65"),
          operatorId = NULL,
          amount = 0L,
          brainAreaIds = list(),
          arbCenter = list(x = NULL, y = NULL, z = NULL),
          arbSize = NULL,
          invert = FALSE,
          composition = 3L
        )
      )
    )),
    query=query
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res <-  mouselight_fetch(path = "graphql",
                           body = bodyj,
                           parse.json = TRUE,
                           simplifyVector=TRUE,
                           include_headers = FALSE,
                           config = httr::content_type_json(),
                           encode='raw')
  som.df = data.frame(idString = res$data$searchNeurons$neurons[[2]],
                      neuron.id = res$data$searchNeurons$neurons[[1]],
                      do.call(rbind, lapply(res$data$searchNeurons$neurons[[3]], function(x) unlist(nullToNA(x)))))
  rownames(som.df) = som.df$idString
  som.df
}


# in progress
# @param structure when searching for the brain regions containing neurons of interest,
# the part of the neuron that should be used for the search
mouselight_neurons_by_brain_region <- function(structure = c("either","axon", "dendrite")){
  structure <- match.arg(structure)
  body <-  list(
    operationName = "SearchNeurons",
    variables = list(context = list(
      scope = 6L,
      nonce = "cjyzvdm9400053h5y90ywg53t",
      predicates = list(
        list(
          predicateType = 3L,
          tracingIdsOrDOIs = list("1"),
          tracingIdsOrDOIsExactMatch = FALSE,
          tracingStructureIds = list("aef2ba31-8f9b-4a47-9de0-58dab1cc06a8"), # .structureIdentifiers$id[.structureIdentifiers$name==structure]
          nodeStructureIds = list(), # end points: "c37953e1-a1e9-4b9a-847e-08d9566ced65"
          operatorId = "f191e8b3-8fb9-4151-a48c-432c1a2382cd",
          amount = 0L,
          brainAreaIds = list("825cacf6-2a60-4fd9-b749-414bd6941357"),
          arbCenter = list(x = NULL, y = NULL, z = NULL),
          arbSize = NULL,
          invert = FALSE,
          composition = 2L
        )
      )
    )),
    query= "query SearchNeurons($context: SearchContext)\n
                  {\n
                    searchNeurons(context: $context) {\n
                    neurons {\n
                    id\n
                    __typename\n
                    brainArea {\n
                              acronym\n
                              safeName\n
                              structureId\n
                            }\n
                    }\n
                  }\n
              }"
    )
  bodyj <- jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res <-  mouselight_fetch(path = "graphql",
                         body = bodyj,
                         parse.json = TRUE,
                         simplifyVector=FALSE,
                         include_headers = FALSE,
                         config = httr::content_type_json(),
                         encode='raw')
  none <-  unlist(lapply(res$data$searchNeurons$neurons, function(x) !is.null(x$brainArea)))
  df <-  data.frame(do.call(rbind, lapply(res$data$searchNeurons$neurons[none], function(x) c(id = x$id, typename = x$`__typename`, unlist(nullToNA(x$brainArea))))))
  rownames(df) <- df$id
  df
}





R/calculate.R
#' Find out how many nodes a neuron has in each brain region
#'
#' @description After reading tracgings from the MouseLight project (\code{\link{mouselight_read_brain}}),
#' find out in which neuropils it has its nodes
#'
#' @param x a \code{neuronlist} of tracings that have been read into R using \code{\link{mouselight_read_neurons}}, with \code{method = "native"}.
#' @param brain.areas a vector of brain region acronyms. See \code{\link{mouselight_brain_region_info}}.
#' @param labels a vector of label IDs, denoting neuron compartment e.g. 2 for axon, 3 for dendrite, 0 for intervening path, etc. See \code{\link{mouselight_read_neurons}}.
#' @param ... methods sent to \code{mouselight_fetch_swc}
#'
#' @details Note, when using \code{mouselight_neurons_by_brain_region}, the brain region
#'   returned can belong to larger brain regions, which are agglomerations of
#'   smaller ones. See \code{\link{mouselight_brain_region_info}}.
#' @return A data.frame
#' @seealso \code{\link{mouselight_read_brain}}, \code{\link{mouselight_brain_region_info}}, \code{\link{mouselight_neuron_info}}, \code{\link{mouselight_read_neurons}}
#'
#' @examples
#' \donttest{
#' ids=c("1ccbc478-88fb-4181-a4de-d0c22ed9738b",
#'   "3afa5e41-374d-45bc-a546-03f362d93649")
#' nn=mouselight_read_neurons(ids, method = "native")
#' nibr = mouselight_nodes_in_region(nn, labels = c(0, 2, 3))
#' print(nibr)
#'
#' # Counts in a specific brain area
#' cing.nodes = mouselight_nodes_in_region(nn, labels = c(0, 2, 3), brain.areas = "cing")
#' print(cing.nodes)
#'
#' }
#' @export
#' @rdname mouselight_nodes_in_region
mouselight_nodes_in_region <-function(x, brain.areas = NULL, labels = NULL) UseMethod("mouselight_nodes_in_region")

#' @export
#' @rdname mouselight_nodes_in_region
mouselight_nodes_in_region.neuron <- function(x, brain.areas = NULL, labels = NULL){
  if(!is.null(labels)){
    points = x$d[x$d$Label %in% labels,]
  }else{
    points = x$d
  }
  if(nrow(points)==0){
    return(0)
  }
  if(!is.null(brain.areas)){
    sum(points$brain_area%in%brain.areas)
  }else{
    counts = c(table(points$brain_area))
    as.data.frame((t(data.frame(counts))))
  }
}

#' @export
#' @rdname mouselight_nodes_in_region
mouselight_nodes_in_region.neuronlist <- function(x, brain.areas = NULL, labels = NULL){
  if(!is.null(brain.areas)){
    sapply(x, FUN = mouselight_nodes_in_region.neuron, brain.areas = brain.areas, labels = labels)
  }else{
    l = lapply(x, FUN = mouselight_nodes_in_region.neuron, brain.areas = brain.areas, labels = labels)
    l = nullToNA(l)
    l = do.call(plyr::rbind.fill, l)
    l[is.na(l)] = 0
    rownames(l) = names(x)
    l
  }
}


R/get_meshes.R
#' @title Read the brain mesh and its brain regions from the MouseLight project
#'
#' @description Save  the brain mesh and its brain regions from the MouseLight project
#' @param type whether just to retreive the outer brain mesh, or the brain constituent neuropils
#' @param progress logical, toggle progress bar
#' @param Force logical, whether or not to overwrite .obj files stored in a temporary directory
#' @param ... methods passed to \code{mouselight_fetch}
#' @return a \code{nat} package \code{\link[nat]{hxsurf}} object, which mimics the Amira surface format, replete with metadata that can be
#' accessed using \code{$}
#' @seealso \code{\link{mouselight_read_brain}}, \code{\link{mouselight_brain_region_info}}
#' @examples
#' \dontrun{
#' ## First we can quickly just plot the outer mesh for the brain
#' outline = mouselight_read_brain(type = "outline")
#' plot3d(outline, col = "pink", alpha = 0.3)
#'
#' ## This is cool, but maybe what we really want are its sub-divisions.
#' mousebrain = mouselight_read_brain(type = "brain_areas")
#' clear3d()
#' plot3d(mousebrain)
#' ### This takes a long time the first time you call this function per session
#'
#' ## What brain regions are on offer?
#' print(mousebrain$neuropil_full_names)
#'
#' ## Or if we want more information, we can get it like this:
#' mbr = mouselight_brain_region_info()
#' View(mouselight_brain_region_info)
#'
#' ## Perhaps we want to plot just the amygdala?
#' ### To do this we can do
#' amygdala.codes = mousebrain$RegionList[grepl("amygdala",
#' mousebrain$neuropil_full_names,
#' ignore.case = TRUE)]
#' plot3d(outline, col = "pink", alpha = 0.1)
#' plot3d(subset(mousebrain, amygdala.codes), alpha = 0.5)
#'
#' }
#' @references Winnubst, Johan, Erhan Bas, Tiago A. Ferreira, Zhuhao Wu, Michael N. Economo, Patrick Edson, Ben J. Arthur, et al. 2019. “Reconstruction of 1,000 Projection Neurons Reveals New Cell Types and Organization of Long-Range Connectivity in the Mouse Brain.” bioRxiv. https://doi.org/10.1101/537233.
#' Economo, Michael N., Nathan G. Clack, Luke D. Lavis, Charles R. Gerfen, Karel Svoboda, Eugene W. Myers, and Jayaram Chandrashekar. 2016. “A Platform for Brain-Wide Imaging and Reconstruction of Individual Neurons.” eLife 5 (January): e10566.
#' @export
#' @rdname mouselight_read_brain
mouselight_read_brain <- function(type = c("outline", "brain_areas"),
                          progress  = TRUE,
                          Force  = FALSE){
  type = match.arg(type)
  temp = tempdir()
  body = list(
    query = "{\n
    brainAreas {\n
    acronym\n
    safeName\n
    geometryFile\n
    }\n
    }",
  variables = list(),
  operationName = NULL
  )
  bodyj=jsonlite::toJSON(body, null = 'null', auto_unbox = T)
  res = mouselight_fetch(path = "graphql",
                 body = bodyj,
                 parse.json = TRUE,
                 simplifyVector=FALSE,
                 include_headers = FALSE,
                 config = httr::content_type_json(),
                 encode='raw')
  df = do.call(rbind,res$data$brainAreas)
  df = as.data.frame(t(apply(df, 1, unlist)))
  df = as.data.frame(df)
  if(type=="outline"){
    df = df[df$safeName == "Whole Brain",]
  }else{
    df = df[df$safeName != "Whole Brain",]
  }
  # Now read .obj files
  urls = mouselight_url("static/allen/obj",df$geometryFile)
  saved.files = success = c()
  for (url in 1:length(urls)){
    message = paste("downloading", df$safeName[url],".obj file", sep = " ")
    localfile = paste0(temp,"/",df$geometryFile[url])
    if(!file.exists(localfile) | Force){
      t = suppressWarnings (try( utils::download.file(urls[url], localfile, mode='wb', quiet = TRUE), silent = TRUE))
      if(inherits(t,'try-error')) {
        message = c("NOTE: unable to download ", df$safeName[url])
        next
      }
    }
    success = c(success, url)
    saved.files = c(saved.files, localfile)
    if(progress) mouselight_progress(url/length(urls)*100, max = 101, message = message)
  }
  objs = lapply(saved.files, function(x)
    tryCatch(readobj::read.obj(x, convert.rgl = T), error = function(e) NULL)[[1]])
  if(progress) mouselight_progress(url/length(urls)*100, max = 101, message = "reading .obj files")
  success = success[!sapply(objs, is.null)]
  objs = objs[!sapply(objs, is.null)]
  objs = lapply(objs, nat::as.hxsurf)
  obj.name = df$acronym
  # build single hxsurf object
  brain = list()
  brain$Vertices = data.frame()
  brain$Regions = list()
  brain$RegionList = brain$RegionColourList = c()
  class(brain) = c("hxsurf","list")
  count = 0
  structure.colors = grDevices::rainbow(length(success))
  for(i in 1:length(objs)){
    o = objs[[i]]
    v = o$Vertices
    v$PointNo = v$PointNo + count
    brain$Vertices = rbind(brain$Vertices, v)
    region.name = obj.name[success][i]
    regions = o$Regions$Interior+count
    brain$Regions[[region.name]] = regions
    brain$RegionList = c(brain$RegionList, region.name)
    brain$RegionColourList = c(brain$RegionColourList, structure.colors[success][i])
    count = count + nrow(o$Vertices)
    if(progress) mouselight_progress(i/length(objs)*100, max = 100, message = "assembling hxsurf object")
  }
  brain$neuropil_full_names = df$safeName[success]
  brain
}

